packages/core/react/src/WalletProvider.tsx

```diff
@@ -23,285 +24,158 @@ export interface WalletProviderProps {
-            isDisconnecting.current = false;
+        function handleBeforeUnload() {
+            isUnloading.current = true;
         }
-    }, [isDisconnecting, adapter, setName]);
-
-    // Send a transaction using the provided connection
-    const sendTransaction: WalletAdapterProps['sendTransaction'] = useCallback(
-        async (transaction, connection, options) => {
-            if (!adapter) throw handleError(new WalletNotSelectedError());
-            if (!connected) throw handleError(new WalletNotConnectedError());
-            return await adapter.sendTransaction(transaction, connection, options);
-        },
-        [adapter, handleError, connected]
-    );
-
-    // Sign a transaction if the wallet supports it
-    const signTransaction: SignerWalletAdapterProps['signTransaction'] | undefined = useMemo(
-        () =>
-            adapter && 'signTransaction' in adapter
-                ? async (transaction) => {
-                      if (!connected) throw handleError(new WalletNotConnectedError());
-                      return await adapter.signTransaction(transaction);
-                  }
-                : undefined,
-        [adapter, handleError, connected]
-    );
-
-    // Sign multiple transactions if the wallet supports it
-    const signAllTransactions: SignerWalletAdapterProps['signAllTransactions'] | undefined = useMemo(
-        () =>
-            adapter && 'signAllTransactions' in adapter
-                ? async (transactions) => {
-                      if (!connected) throw handleError(new WalletNotConnectedError());
-                      return await adapter.signAllTransactions(transactions);
-                  }
-                : undefined,
-        [adapter, handleError, connected]
-    );
-
-    // Sign an arbitrary message if the wallet supports it
-    const signMessage: MessageSignerWalletAdapterProps['signMessage'] | undefined = useMemo(
-        () =>
-            adapter && 'signMessage' in adapter
-                ? async (message) => {
-                      if (!connected) throw handleError(new WalletNotConnectedError());
-                      return await adapter.signMessage(message);
-                  }
-                : undefined,
-        [adapter, handleError, connected]
-    );
-
+        /**
+         * Some wallets fire disconnection events when the window unloads. Since there's no way to
+         * distinguish between a disconnection event received because a user initiated it, and one
+         * that was received because they've closed the window, we have to track window unload
+         * events themselves. Downstream components use this information to decide whether to act
+         * upon or drop wallet events and errors.
+         */
+        window.addEventListener('beforeunload', handleBeforeUnload);
+        return () => {
+            window.removeEventListener('beforeunload', handleBeforeUnload);
+        };
+    }, [adapters, walletName]);
```

Comment:
Are there other wallets besides Phantom that we should request to fix this behavior so we can remove this hack that hurts cache perf?

Reply:
Super good question. We could do an audit, but we could also just wait until every wallet supports the standard and this doesn't matter, right? Or does the standard not help with this?

Reply:
Hmm, the standard doesn't help with this _yet_ because when used as a wrapper, it's still just downstream of `window.phantom` or whatever is firing disconnect events. I think we fix it in Phantom, make sure that Glow, Solflare, Backpack aren't affected, then we burn it (tracking issue for completion https://github.com/solana-labs/wallet-adapter/issues/593)